text/0000-pointer-match.md

```diff
@@ -0,0 +1,329 @@
+  possible that not a single variant is currently completely initialized, yet
+  one wants to access some subfield. See
+  <https://github.com/rust-lang/unsafe-code-guidelines/issues/73#issuecomment-460634637>.
+* Doing pointer offset calculations where the references does not refer to the
+  same, or any, allocation. This is because reference calculations are
+  performed with `getelementptr inbounds`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Match expression are extended from support for a reference binding mode, to a
+pointer binding mode. Furthermore, a new pattern binds to a pointer, and
+identifiers are extended to allow a new mode similar to `ref` and `ref mut`
+binding to a reference. These patterns are called pointer pattern and raw
+identifier for the remainder of the document.
+
+```
+#[repr(packed)]
+struct Foo {
+    a: u16,
+    b: u32,
+}
+
+fn ptr_b(foo: &mut Foo) -> *mut u32 {
+    let Foo { raw mut b, .. } = foo;
+    b
+}
+```
+
+Note that pointer binding mode and pointer pattern requires `unsafe`, even when
+it will never dereference the pointer. But the arithmetic on the pointer may
+implicitely overflow. Furthermore, not all patterns are (yet) allowed, to avoid
+implicitely performing an unintended, unsafe read through the pointer. Pointer
+binding mode will at first only permit ultimately binding with `raw` and `ref`
+and not actually reading the contained memory.
+
+The raw identifier pattern does not require `unsafe` on its own (as seen above,
+where we safely match a `&mut` but bind to `*mut`).
+
+This is not only useful for packed fields, but also to access the fields of
+*any* object that is only available via pointer because its state invariants
+may not yet be fulfilled. Instead of manually doing pointer math:
+
+```
+/// Repr Rust, so no layout guarantees, no pointer operations to get to `a`.
+struct Weird {
+    /// Always valid, no matter the memory content.
+    something_i_dont_care_about: u8,
+
+    /// Must only be one of `true` and `false` for a &Weird.
+    a: bool,
+}
+
+/// Unsafety invariants: `w` must 
+/// * point to some allocation of at least `std::mem::size_of::<Weird>`.
+/// * point to memory valid for the chosen lifetime `'a`
+/// * be properly aligned.
+unsafe fn get_if_init<'a>(w: *const Weird) -> Option<&'a Weird> {
+    let Weird { raw const a, ..} = w;
+    match core::ptr::read(a as *const u8) {
+    	0 | 1 => Some(std::mem::transmute(w)),
+	_ => None
+    }
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The newly introduced patterns are:
+
+* `raw (const|mut) identifier`; allowed for field bindings and identifier bindings.
+  These are allowed in the grammar where `ref? mut?  identifier` is allowed
+  currently. For this purpose `raw` is a contextual keyword.
+* `* <subpattern>`; to match a pointer not by value but to additionally use
```

Comment:
Is there a reason to have this be a generic `*` pattern instead of separate `*const`, `*mut` patterns?

Reply:
Not really, and it does seem better to stay consistent with `&`/`&mut.

Reply:
Using `*const`/`*mut` isn't really consistent with `&`/`&mut` in a meaningful way, AFAICS. Patterns follow their corresponding expression syntax, and while the way to construct a `&T`/`&mut T` is `&x`/`&mut x`, the way to construct a `*const T` is not `*const x`.

Reply:
So.....

```
match (0 as *mut usize) {
    &mut raw const z as *mut usize => (),
}
```

(definitely not serious).

---

It occurs to me that I mostly think about patterns following the _type_ syntax rather than expression syntax, which is likely why I made this comment. It just works out that those are the same most of the time because of struct expressions following the type syntax.

Reply:
Agree with @Nemo157 here, I had always assumed it followed type syntax due to `struct`-patterns and `enum`-patterns even requiring the explicit name and e.g. not accepting `Self { }`